iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp

```diff
@@ -397,6 +497,60 @@ struct PromoteConvolutionSubviewsPattern
 };
 }  // namespace
 
+/// Apply canonicalizations related to tiling to make promotion/vectorization
+/// easier.
+static void applyCanonicalizationPatterns(MLIRContext *context, Operation *op) {
+  OwningRewritePatternList canonicalizationPatterns;
+  canonicalizationPatterns.insert<AffineMinCanonicalizationPattern>(context);
+  AffineApplyOp::getCanonicalizationPatterns(canonicalizationPatterns, context);
+  AffineMinOp::getCanonicalizationPatterns(canonicalizationPatterns, context);
+  SubViewOp::getCanonicalizationPatterns(canonicalizationPatterns, context);
+  applyPatternsAndFoldGreedily(op, canonicalizationPatterns);
+}
+
+/// Apply second-level of tiling patterns to enable targeting vectorization.
+static void applySubgroupTiling(MLIRContext *context, Operation *op,
+                                TileSizeCalculator const &tileSizeCalculator) {
+  OwningRewritePatternList subgroupPatterns;
+  linalg::LinalgTilingOptions tilingOptions;
+  tilingOptions.setLoopType(linalg::LinalgTilingLoopType::ParallelLoops);
+
+  auto insertMatmulPattern = [&](Type dataType, Type accumulatorType) {
+    ArrayRef<int64_t> tileSizes = tileSizeCalculator.getTileSizes();
+    ArrayRef<int64_t> workgroupSize = tileSizeCalculator.getWorkgroupSize();
+    if (tileSizes.size() < 2 || workgroupSize.size() < 2) return;
+    Optional<std::array<int64_t, 3>> subgroupTileSizes =
+        getCooperativeMatmulSubgroupSize(dataType, accumulatorType);
+    if (!subgroupTileSizes || tileSizes[0] % (*subgroupTileSizes)[0] ||
+        tileSizes[1] % (*subgroupTileSizes)[1])
+      return;
+    std::array<int64_t, 2> numSubgroups = {
+        std::min(tileSizes[1] / (*subgroupTileSizes)[1], workgroupSize[0]),
+        std::min(tileSizes[0] / (*subgroupTileSizes)[0], workgroupSize[1])};
```

Comment:
Why the number of subgroups is related to the tile size? It should be only dependent on the workgroup size and subgroupSize.